lib/nrf_modem_lib/trace_backends/uart/uart.c

```diff
@@ -1,142 +1,172 @@
+	ARG_UNUSED(dev);
+	ARG_UNUSED(user_data);
 
+	switch (evt->type) {
+	case UART_TX_DONE:
+		tx_succeeded = true;
 		k_sem_give(&tx_sem);
-	}
-
-	if (event->type == NRFX_UARTE_EVT_TX_DONE) {
+		break;
+	case UART_TX_ABORTED:
+		tx_succeeded = false;
 		k_sem_give(&tx_sem);
+		break;
+	default:
+		LOG_DBG("Unhandled UART event: %d", evt->type);
+		break;
 	}
 }
 
-static void wait_for_tx_done(void)
-{
-	/* Attempt to take the TX semaphore to stop the thread execution until
-	 * UART is done sending. Immediately give the semaphore when it becomes
-	 * available.
-	 */
-	if (k_sem_take(&tx_sem, K_MSEC(UART_TX_WAIT_TIME_MS)) != 0) {
-		LOG_WRN("UARTE TX not available");
-	}
-
-	k_sem_give(&tx_sem);
-}
-
 int trace_backend_init(trace_backend_processed_cb trace_processed_cb)
 {
 	int err;
-	const nrfx_uarte_config_t config = {
-		.skip_gpio_cfg = true,
-		.skip_psel_cfg = true,
-		.hal_cfg.hwfc = NRF_UARTE_HWFC_DISABLED,
-		.hal_cfg.parity = NRF_UARTE_PARITY_EXCLUDED,
-		.baudrate = NRF_UARTE_BAUDRATE_1000000,
-		.interrupt_priority = DT_IRQ(UART1_NL, priority),
-		.p_context = NULL,
-	};
+	struct uart_config uart_cfg;
 
 	if (trace_processed_cb == NULL) {
 		return -EFAULT;
 	}
 
 	trace_processed_callback = trace_processed_cb;
 
-	err = pinctrl_apply_state(PINCTRL_DT_DEV_CONFIG_GET(UART1_NL), PINCTRL_STATE_DEFAULT);
-	__ASSERT_NO_MSG(err == 0);
+	if (!device_is_ready(uart_dev)) {
+		printk("UART device not found!");
+		return -ENODEV;
+	}
+
+	err = uart_config_get(uart_dev, &uart_cfg);
+	if (err) {
+		LOG_ERR("UART_config_get failed: %d", err);
+		return err;
+	}
 
-	IRQ_CONNECT(DT_IRQN(UART1_NL), DT_IRQ(UART1_NL, priority), nrfx_isr,
-		    &nrfx_uarte_1_irq_handler, UNUSED_FLAGS);
+	if (uart_cfg.baudrate != TRACE_UART_BAUD_REQUIRED) {
+		LOG_ERR("Trace UART baudrate is %d but %d is required",
+			uart_cfg.baudrate, TRACE_UART_BAUD_REQUIRED);
+		return -ENOTSUP;
+	}
 
-	err = nrfx_uarte_init(&uarte_inst, &config, &uarte_callback);
-	if (err != NRFX_SUCCESS && err != NRFX_ERROR_INVALID_STATE) {
-		return -EBUSY;
+	err = uart_err_check(uart_dev);
+	if (err) {
+		LOG_ERR("UART check failed: %d", err);
+		return -EIO;
+	}
+
+	err = uart_callback_set(uart_dev, uart_callback, NULL);
+	if (err) {
+		LOG_ERR("Cannot set callback: %d", err);
+		return -EFAULT;
 	}
 
 	return 0;
 }
 
 int trace_backend_deinit(void)
 {
-	nrfx_uarte_uninit(&uarte_inst);
+	return 0;
+}
 
-#if CONFIG_PM_DEVICE
+/* Returns the number of bytes written, or negative error. */
+static int uart_send(const uint8_t *data, size_t len)
+{
 	int err;
 
-	err = pinctrl_apply_state(PINCTRL_DT_DEV_CONFIG_GET(UART1_NL), PINCTRL_STATE_SLEEP);
-	__ASSERT_NO_MSG(err == 0);
-#endif /* CONFIG_PM_DEVICE */
+	for (int i = 0; i < UART_TX_RETRIES; i++) {
+		err = uart_tx(uart_dev, data, len,
+			UART_TX_WAIT_TIME_MS * USEC_PER_MSEC);
+		if (err) {
+			LOG_ERR("uart error: %d", err);
+			return err;
+		}
+
+		err = k_sem_take(&tx_sem, K_FOREVER);
```

Comment:
With infinite timeout, I don't think `k_sem_take()` returns unless `k_sem_reset()` is called or the semaphore is successfully taken. I don't see `k_sem_reset()` being called, so the return value check is probably not needed?